### PR TITLE
Fix learning rate schedule in Megatron models when `max_steps` is not set

### PR DESCRIPTION
**Pull request by odelalleau**
# What does this PR do ?

This fixes the learning rate schedule in Megatron models when relying on `max_epochs` instead of `max_steps`.

The issue is that [compute_max_steps](https://github.com/NVIDIA/NeMo/blob/bdea326811ef81d3aff3530b26b67f7a7ce5119c/nemo/core/optim/lr_scheduler.py#L934) is incorrect for Megatron models since `accumulate_grad_batches` is equal to 1 from PTL's perspective, even if we are actually using multiple micro-batches.

**Collection**: nlp

# Changelog 
- Fix learning rate schedule in Megatron models when `max_steps` is not set

# Before your PR is "Ready for review"
**Pre checks**:
- [X] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [X] Bugfix
- [ ] Documentation

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

I considered two potential implementations:

1. (this one) setting `max_steps` directly in `MegatronBaseModel`
2. Setting `t_accumulate_grad_batches` correctly in the scheduler config so that the formula is correct

I didn't go with (2) because it seemed trickier to ensure that all parameters are correct so that the formula works, including with drop_last which [still raises an old warning](https://github.com/NVIDIA/NeMo/blob/a420f9083ea9993152a5221b8875c34470561f13/nemo/core/optim/lr_scheduler.py#L943). It's easier to just compute `max_steps` within `MegatronBaseModel`.

cc @aklife97

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_base_model.py
Comment (odelalleau) [inline nemo/collections/nlp/models/language_modeling/megatron_base_model.py:453]: This creates a copy of the config, and another one is also created in `ModelPT` ==> we end up copying it twice. It's a bit unsatisfying but since it's a very small config the performance hit is negligible, and trying to avoid the extra copy would probably make the code more compelx.

File: nemo/core/classes/modelPT.py
Comment (odelalleau) [inline nemo/core/classes/modelPT.py:617]: I simplified a bit this code -- for instance the `with open_dict(self._cfg.optim)` was not needed since we are setting `self._cfg.optim` directly (we're not modifying its attributes).

File: nemo/core/classes/modelPT.py
Comment (titu1994) [inline nemo/core/classes/modelPT.py:1746]: Why not just "_optim_copy_config"

**Review by titu1994 — COMMENTED** [review]
Overall logic seems reasonable, let's let tests pass

File: nemo/core/classes/modelPT.py
Comment (odelalleau) [reply nemo/core/classes/modelPT.py:1746]: Happy to rename it but did you really mean `_optim_copy_config()`, or would it be `_optim_config_copy()` / `_copy_optim_config()`?

Comment (titu1994) [reply nemo/core/classes/modelPT.py:1746]: _optim_config_copy is better

Comment (odelalleau) [reply nemo/core/classes/modelPT.py]: Sounds good, done in 8e556a9dc

**Comment by odelalleau** [comment]
_Commenting to un-stale PR_

File: nemo/collections/nlp/models/language_modeling/megatron_base_model.py
Comment (titu1994) [inline nemo/collections/nlp/models/language_modeling/megatron_base_model.py:536]: Why should this be set manually on the config above ? Trainer has this value

File: nemo/collections/nlp/models/language_modeling/megatron_base_model.py
Comment (titu1994) [inline nemo/collections/nlp/models/language_modeling/megatron_base_model.py:867]: Don't use asserts everywhere, check if trainer is none or not and fall back. It's not guaranteed to have trainer when building a non LLM model, but setup optimization and gets called on model restore when there is no trainer. So all these asserts raise the possibility of breaking model restore in untested cases.

**Review by titu1994 — COMMENTED** [review]
The logic seems fine, but I dunno about the asserts. Core changes need 2+ reviews, so @ericharper if you are ok with this change in nlp then let's merge

File: nemo/collections/nlp/models/language_modeling/megatron_base_model.py
Comment (odelalleau) [reply nemo/collections/nlp/models/language_modeling/megatron_base_model.py:536]: The whole point of this PR is to fix the situation where `trainer.max_steps` is `None`. We need `cfg.optim.sched.max_steps` to be set for the scheduler to work, so we use `trainer.max_epochs` to compute `max_steps` in this situation (in the new `_get_max_steps()` function added in this PR).

This assert is a sanity check, it shouldn't possible for it to fail in the current code since we do `optim_config.sched.max_steps = self._get_max_steps()`. I added it to (i) make sure my code is working as intended, and (ii) make sure no-one breaks it in the future.

File: nemo/collections/nlp/models/language_modeling/megatron_base_model.py
Comment (odelalleau) [reply nemo/collections/nlp/models/language_modeling/megatron_base_model.py:867]: Ah thanks for pointing this out (though just a note that this is in `MegatronBaseModel` in nlp collection, so this shouldn't affect non-LLM models, right?)

I pushed a tentative fix in 4c991f553. Will mark this PR as draft while I'm testing it.

File: nemo/collections/nlp/models/language_modeling/megatron_base_model.py
Comment (odelalleau) [reply nemo/collections/nlp/models/language_modeling/megatron_base_model.py]: > oh, I just noticed the assert self._trainer.max_steps is not None above... now I'm a bit confused as to how I was testing this, it's been a long time and I don't remember :/

After testing it I now remember: it's because `max_steps` is set to -1 when undefined (which is confusing because `max_epochs` is using `None`). I think this assert was actually a mistake, assuming that a `None` value could appear when undefined, but I tested setting `max_steps: null` in the config and it actually crashes earlier during PTL trainer initialization because `None` is not valid.

=> I removed this assert in ff09438c5 (and added another one that validates that nobody uses a negative value for `max_steps` in the scheduler config, since we use `None` here for undefined).

File: nemo/collections/nlp/models/language_modeling/megatron_base_model.py
Comment (odelalleau) [reply nemo/collections/nlp/models/language_modeling/megatron_base_model.py]: > I pushed a tentative fix https://github.com/NVIDIA/NeMo/commit/3399ce842046c18e5852a075f1b457e426a8ff52. Will mark this PR as draft while I'm testing it.

Testing it made me realize I didn't remember correctly the code flow when writing this fix, and now I remember why I was triggering exceptions in `_get_max_steps()` if it couldn't be computed: this is because if we don't set `max_steps` properly, we fall back to the original issue where it's computed incorrectly.

One reason for my confusion (which I'm mentioning to be sure it doesn't confuse you as well) is that the `prepare_lr_scheduler()` call in `MegatronBaseModel.configure_optimizers()` ([here](https://github.com/NVIDIA/NeMo/blob/fd8f234070b21c64191824bf6b0f3d0cbdc2b61b/nemo/collections/nlp/models/language_modeling/megatron_base_model.py#L531)) is not actually the "usual" place where the scheduler is configured (this code is guarded by an `if` block) -- usually this happens in `ModelPT.setup_optimization()` ([here](https://github.com/NVIDIA/NeMo/blob/fd8f234070b21c64191824bf6b0f3d0cbdc2b61b/nemo/core/classes/modelPT.py#L733)).

So if we don't crash in `_get_max_steps()` (when we can't compute it), since we don't usually go through the check I added in `MegatronBaseModel.configure_optimizers()`, then we end up using a wrong schedule (same issue that this PR is trying to fix).

I pushed a new version in a26ac7535. This one actually fully relaxes the constraint that `max_steps` must be computed, and just skips the scheduler if it can't. This seems a bit dangerous to me though and I wanted to ask you if you think it'd be fine to trigger an exception in `MegatronBaseModel.configure_optimizers()` (not `setup_optimization()`), or if we also need that one to work even when the model is not fully setup for training?

**Comment by odelalleau** [comment]
Just realized I had forgotten to un-draft this PR... should be ready for a second look now!

**Comment by odelalleau** [comment]
Re-opening as this bit us again.

@titu1994 would you be able to have another look? (let me know if anything's still unclear, I realize there was some back and forth on some points that may have been confusing)

**Comment by odelalleau** [comment]
jenkins

**Comment by odelalleau** [comment]
@titu1994 would you be able to help review this one? Thanks!

File: nemo/collections/nlp/models/language_modeling/megatron_base_model.py
Comment (titu1994) [reply nemo/collections/nlp/models/language_modeling/megatron_base_model.py]: Silent about scheduler - it was a (bad) default in NeMo but it's easily detected by looking at logs.
For the megatron case, fine to assert for nlp only, other domains let them continue on without scheduler if max steps cannot be calculated (old default)

**Comment by titu1994** [comment]
jenkins

**Comment by titu1994** [comment]
Jenkins seems to be failing but overall this is fine. Core changes need two reviews, so @ericharper for second review

File: nemo/collections/nlp/models/language_modeling/megatron_base_model.py
Comment (odelalleau) [reply nemo/collections/nlp/models/language_modeling/megatron_base_model.py]: > For the megatron case, fine to assert for nlp only, other domains let them continue on without scheduler if max steps cannot be calculated (old default)

Sounds good, I added a check in b6354f120 to raise an exception when the scheduler can't be instantiated in Megatron.

**Comment by odelalleau** [comment]
jenkins

**Comment by odelalleau** [comment]
> Jenkins seems to be failing

I'll have a look after it re-runs on the latest version.

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

Ran nightly nemo-ci tests as well and all tests passed.

File: nemo/core/optim/lr_scheduler.py
Comment (stevehuang52) [inline nemo/core/optim/lr_scheduler.py:837]: what should we do if we want epoch based training? by default `max_steps` is set to -1, which means it'll return here before calculating the actual max_steps later in this function. `scheduler_config.get('max_steps')` will never be None, since otherwise PTL will throw an error saying max_teps must be an integer. @odelalleau @titu1994 @ericharper

**Pull request merged**
